src/dex/idle-dao/idle-dao.ts

```diff
@@ -0,0 +1,372 @@
+      });
+
+      return [
+        {
+          unit: prices[0],
+          prices: prices.slice(1),
+          gasCost: this.config.lendingGasCost,
+          exchange: this.dexKey,
+          data: {
+            idleToken: {
+              cdoAddress: idleToken.cdoAddress,
+              tokenType: idleToken.tokenType,
+            },
+            fromIdleToken,
+          },
+          poolAddresses: [fromIdleToken ? srcToken.address : destToken.address],
+        },
+      ];
+    } catch (e) {
+      if (blockNumber === 0)
+        this.logger.error(
+          `Error_getPricesVolume: Aurelius block manager not yet instantiated`,
+        );
+      this.logger.error(`Error_getPrices:`, e);
+      return null;
+    }
+  }
+
+  // Cannot be implemented fully event-based.
+  // Rate changes almost every block.
+  // rpc-based pricing is only done for limited amount of tokens that are supported
+  async getTokenPrice(
+    idleToken: IdleToken,
+    blockNumber: number,
+  ): Promise<bigint | null> {
+    if (!idleToken.cdoContract) {
+      return null;
+    }
+
+    const tranchePrice = await idleToken.cdoContract.methods['virtualPrice'](
+      idleToken.idleAddress,
+    ).call({}, blockNumber);
+
+    return BigInt(tranchePrice);
+  }
+
+  // Returns estimated gas cost of calldata for this DEX in multiSwap
+  getCalldataGasCost(poolPrices: PoolPrices<IdleDaoData>): number | number[] {
+    return CALLDATA_GAS_COST.DEX_NO_PAYLOAD;
+  }
+
+  getAdapterParam(
+    srcToken: string,
+    destToken: string,
+    srcAmount: string,
+    destAmount: string,
+    data: IdleDaoData,
+    side: SwapSide,
+  ): AdapterExchangeParam {
+    return {
+      // target exchange is not used by the contract
+      targetExchange: NULL_ADDRESS,
+      payload: '0x',
+      networkFee: '0',
+    };
+  }
+
+  getDexParam(
+    srcToken: Address,
+    destToken: Address,
+    srcAmount: NumberAsString,
+    destAmount: NumberAsString,
+    recipient: Address,
+    data: IdleDaoData,
+    side: SwapSide,
+    _: Context,
+    executorAddress: Address,
+  ): DexExchangeParam {
+    let returnAmountPos = undefined;
+
+    const [Interface, swapCallee, swapFunction, swapFunctionParams] = ((): [
+      Interface,
+      Address,
+      PoolFunctions,
+      Param,
+    ] => {
+      if (data.fromIdleToken) {
+        returnAmountPos = extractReturnAmountPosition(
+          this.cdo,
+          PoolFunctions[`withdraw${data.idleToken.tokenType}`],
+        );
+        return [
+          this.cdo,
+          data.idleToken.cdoAddress,
+          PoolFunctions[`withdraw${data.idleToken.tokenType}`],
+          [srcAmount],
+        ];
+      }
+
+      returnAmountPos = extractReturnAmountPosition(
+        this.cdo,
+        PoolFunctions[`deposit${data.idleToken.tokenType}`],
+      );
+      return [
+        this.cdo,
+        data.idleToken.cdoAddress,
+        PoolFunctions[`deposit${data.idleToken.tokenType}`],
+        [srcAmount, recipient],
```

Comment:
why are we passing recipient as a referral? and why are we using deposit with referral in the first place?

Reply:
it's been done by the dex team, don't see any issue with that, maybe want to track swaps

Reply:
the only issue is that it is a bit more expensive, but its fine if the dex team has reasons for using it